Source/WebCore/page/Page.cpp

```diff
@@ -2494,16 +2494,20 @@ void Page::stopKeyboardScrollAnimation()
         if (!frameView)
             continue;
 
-        frameView->stopKeyboardScrollAnimation();
-
         auto scrollableAreas = frameView->scrollableAreas();
         if (!scrollableAreas)
             continue;
 
         for (auto& scrollableArea : *scrollableAreas) {
             // First call stopAsyncAnimatedScroll() to prepare for the keyboard scroller running on the scrolling thread.
             scrollableArea->stopAsyncAnimatedScroll();
-            scrollableArea->stopKeyboardScrollAnimation();
+            scrollableArea->stopKeyboardScrollAnimation([&scrollableArea, &frame] {
+                auto page = frame->page();
+                if (!page)
+                    return;
+                if (auto scrollingCoordinator = frame->page()->scrollingCoordinator())
+                    scrollingCoordinator->setKeyboardScrollAnimationInProgress(*scrollableArea, false);
```

Comment:
This seems racy -- what if you've started another scroll in the meantime?

Reply:
Ah ok. Do you have any ideas on how to make it safer?

Reply:
We usually use incrementing integer request identifiers ("make sure I'm still the latest request to stop, if somebody else came in in the meantime then do nothing")

But it's possible that this thing would be better handled by tokens ("everyone who is doing a keyboard scroll takes out a refcounted token and drops it when the scroll ends, and if there are any outstanding tokens then KeyboardScrollAnimationInProgress == true, and when the last outstanding one dies do whatever you need").

I'm not sure :) Would have to peruse this code a bit more.